src/main/java/com/segment/analytics/android/integrations/firebase/FirebaseIntegration.java

```diff
@@ -3,88 +3,278 @@
+      logger.verbose("firebaseAnalytics.logEvent(%s, %s);", eventName, formattedProperties);
+    } catch (Exception e) {
+      logger.verbose(
+              "You must set `currency` in your event's property object to accurately "
+                      + "pass 'value' to Firebase. This event was not sent to Firebase.");
     }
+  }
 
-    // mFirebaseAnalytics.setUserProperty(key, value);
+  private String mapEvent(String event) {
+    String eventName = event;
+    if (eventMapper.containsKey(eventName)) {
+      eventName = eventMapper.get(eventName);
+    }
+    if (eventName.contains(" ")) {
+      eventName = eventName.trim().toLowerCase().replaceAll(" ", "_");
+    }
+    if (eventName.length() > 40) {
+      eventName = trimKey(eventName);
+    }
+    return eventName;
+  }
 
+  private Bundle formatProperties(Properties properties) throws Exception {
+    if ((properties.revenue() != 0 || properties.total() != 0) && isNullOrEmpty(properties.currency())) {
+      throw new Exception();
+    }
+    Bundle bundle = new Bundle();
+    for (Map.Entry<String, Object> entry : properties.entrySet()) {
+      String property = entry.getKey();
+      property  = mapProperty(property);
+      if (entry.getValue() instanceof Integer) {
+        int value = (int) entry.getValue();
+        bundle.putInt(property, value);
+        logger.verbose("bundle.putInt(%s, %s);", property, value);
+        continue;
+      }
+      if (entry.getValue() instanceof Double) {
+        double value = (double) entry.getValue();
+        bundle.putDouble(property, value);
+        logger.verbose("bundle.putDouble(%s, %s);", property, value);
+        continue;
+      }
+      if (entry.getValue() instanceof Long) {
+        long value = (long) entry.getValue();
+        bundle.putLong(property, value);
+        logger.verbose("bundle.putLong(%s, %s);", property, value);
+        continue;
+      }
+      if (entry.getValue() instanceof String) {
+        String value = String.valueOf(entry.getValue());
+        if (ISO_DATE.matcher(value).matches()) {
+          value = value.substring(0, 10);
+        }
+        bundle.putString(property, value);
+        logger.verbose("bundle.putString(%s, %s);", property, value);
+        continue;
+      }
+      if (entry.getValue() instanceof Date) {
+        Date value = (Date) entry.getValue();
+        String formattedDate = formatDate(value);
+        bundle.putString(property, formattedDate);
+        logger.verbose("bundle.putString(%s, %s);", property, formattedDate);
+        continue;
+      }
+    }
+    return bundle;
   }
 
-  @Override public void screen(ScreenPayload screen) {
-    super.screen(screen);
-//    mFirebaseAnalytics.setCurrentScreen(this, screen.name(), null /* class override */);
+  private String mapProperty(String property) {
+    if (propertyMapper.containsKey(property)) {
+      property = propertyMapper.get(property);
+    }
+    if (property.contains(" ")) {
```

Comment:
The check seems unnecessary. You can do `property = property.trim().replaceAll(" ", "_");` unconditionally, unless it's significantly slower. I think both are O(n), and doing it unconditionally avoids duplicate work.